pkg/bpf/bpffs.go

```diff
@@ -104,100 +118,275 @@ func OpenAfterMount(m *Map) error {
 	return nil
 }
 
+// removeMap removes a map from the BPF filesystem.
+func removeMap(name string) {
+	os.Remove(MapPath(name))
+}
+
+// RemoveAfterMount schedules a map to be deleted after the BPF filesystem has
+// been mounted. If the filesystem is already mounted, the map is deleted
+// immediately.
+func RemoveAfterMount(name string) {
+	mountMutex.Lock()
+	defer mountMutex.Unlock()
+
+	if mounted {
+		removeMap(name)
+	}
+
+	log.WithField(logfields.BPFMapName, name).Debug("bpffs is not mounted yet; adding to list of maps to remove once it is mounted")
+	delayedRemoves = append(delayedRemoves, name)
+}
+
+// mountFS mounts the BPFFS filesystem into the desired mapRoot directory.
 func mountFS() error {
-	// Mount BPF Map directory if not already done
-	args := []string{"-q", mapRoot}
-	_, err := exec.Command("mountpoint", args...).CombinedOutput()
+	mapRootStat, err := os.Stat(mapRoot)
 	if err != nil {
-		args = []string{"bpffs", mapRoot, "-t", "bpf"}
-		out, err := exec.Command("mount", args...).CombinedOutput()
-		if err != nil {
-			return fmt.Errorf("command execution failed: %s\n%s", err, out)
-		}
-	} else { // Already mounted. We need to fail if mounted multiple times.
-
-		// Execute the following command to find multiple bpffs mount points
-		// % mount | grep "<mapRoot> " | wc -l | cut -f1 -d' '
-		newmapRoot := mapRoot + " " // Append space to ignore /sys/fs/bpf/xdp and /sys/fs/bpf/ip mountpoints.
-		cmds := []*exec.Cmd{
-			exec.Command("mount"),
-			exec.Command("grep", newmapRoot),
-			exec.Command("wc", "-l"),
-			exec.Command("cut", "-f1", "-d "),
+		if os.IsNotExist(err) {
+			os.MkdirAll(mapRoot, 0644)
+		} else {
+			return fmt.Errorf("failed to stat the mount path %s: %v", mapRoot, err)
 		}
+	}
 
-		output, stderr, err := pipeexec.CommandPipe(cmds)
-		if len(stderr) > 0 || err != nil {
-			return fmt.Errorf("command execution failed: %s", stderr)
-		}
+	if !mapRootStat.IsDir() {
+		return fmt.Errorf("%s is a file which is not a directory: %v", mapRoot, err)
+	}
 
-		// Strip the newline character at the end.
-		parts := strings.Split(string(output), "\n")
+	if err := syscall.Mount(mapRoot, mapRoot, "bpf", 0, ""); err != nil {
+		return fmt.Errorf("failed to mount %s: %v", mapRoot, err)
+	}
+	return nil
+}
 
-		// Convert the string to integer
-		num, err := strconv.ParseInt(parts[0], 10, 32)
-		if err != nil {
-			return fmt.Errorf("command execution failed: %s", err)
+// mountsScanner is a struct which reads the content of /proc/mounts and allows
+// to read it line by line. Wrapping into a struct simplifies closing the file
+// descriptor.
+type mountsScanner struct {
+	fd      *os.File
+	scanner *bufio.Scanner
+}
+
+func (ms *mountsScanner) Close() error {
+	return ms.fd.Close()
+}
+
+func (ms *mountsScanner) Scan() bool {
+	return ms.scanner.Scan()
+}
+
+func (ms *mountsScanner) Text() string {
+	return ms.scanner.Text()
+}
+
+func newMountsScanner() (*mountsScanner, error) {
```

Comment:
Can you pass the filename as an argument to this function so you can add a unit test?

```
// search and replace newMountsScanner() with mountScanner()
func mountsScanner() (*mountsScanner, error) {
   return newMountsScanner("/proc/mounts")
}

// unit test this function
func newMountsScanner(path string) (*mountsScanner, error) {
....
}
```